### DISCUSSION TITLE
How to store functions in UserData?

### DISCUSSION THREAD

**Discussion by Absolucy** [mlua-rs/mlua#30] [Q&A] [answered]
I'm trying to use `mlua` as a method of my app allowing custom scripts to modify data. However, I cannot figure out how to setup callbacks:

```rust
pub struct Route<'a> {
	callback: LuaFunction<'a>,
}

pub struct Router<'a>(Vec<Route<'a>>);

impl<'a> UserData for Router<'a> {
	fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
		methods.add_method_mut("register", Self::register);
	}
}

impl<'a> Router<'a> {
	fn register(
		_: &Lua,
		router: &mut Router<'a>,
		(_, callback): (LuaValue<'a>, LuaFunction<'a>),
	) -> LuaResult<()> {
		router.0.push(Route {
			callback,
		});
		Ok(())
	}
}
```

This obviously fails, due to lifetime stuff. How would I properly do this?

**Comment by khvzak** [comment]
Could you please provide more context/full example?

Your code compiles fine:
```rust
let mut router = Router(Vec::new());
let f = lua.create_function(|_, ()| Ok(()))?;
Router::register(&lua, &mut router, (LuaNil, f))?;
```

**Reply by Absolucy** [reply]
Whoops, I forgot the actual part that was causing me trouble, the `UserData` implementation.

I updated the original post: 
```rust
impl<'a> UserData for Router<'a> {
	fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
		methods.add_method_mut("register", Self::register);
	}
}
```

**Comment by khvzak** [comment, marked as answer]
Unfortunately you cannot refer to a `LuaFunction` in that way :( 
A possible solution could be storing `Vec<Route>` inside Lua (in a `UserData`) and refer to it using `get_user_value()` / `set_user_value()` or use Registry (via `named_registry_value`).

An example using `get/set_user_value`
```rust
pub struct Router;

impl UserData for Router {
    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_function(
            "register",
            |lua: &Lua, (router, _, callback): (AnyUserData, LuaValue, LuaFunction)| {
                // To get `router` as a Router instance:
                // let router = router.borrow::<Router>()?;
                let routes = match router.get_user_value::<Option<LuaTable>>()? {
                    Some(t) => t,
                    None => {
                        let table = lua.create_table()?;
                        router.set_user_value(table.clone())?;
                        table
                    }
                };
                let len = routes.raw_len();
                routes.raw_insert(len + 1, callback)?;
                Ok(())
            },
        );
    }
}
```

**Reply by Absolucy** [reply]
oh, thanks. although my actual `Route` is a bit more complex, I just simplified it to make it a "minimal working example":
```rust
pub struct Route<'lua> {
	hostnames: Vec<String>,
	matcher: RouteMatcher,
	callback: Function<'lua>,
}

```

**Reply by khvzak** [reply]
You could store callback_idx in this case:
```rust
impl UserData for Router {
    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_function(
            "register",
            |lua: &Lua, (router, _, callback): (AnyUserData, LuaValue, LuaFunction)| {
                let routes = match router.get_user_value::<Option<LuaTable>>()? {
                    Some(t) => t,
                    None => {
                        let table = lua.create_table()?;
                        router.set_user_value(table.clone())?;
                        table
                    }
                };
                let idx = routes.raw_len() + 1;
                routes.raw_insert(idx, callback)?;

                let router = router.borrow_mut::<Router>()?;
                router.0.push(Route {
                    hostnames: Vec::new(),
                    matcher: RouteMatcher,
                    callback_idx: idx,
                });

                Ok(())
            },
        );
    }
}
```

**Reply by Absolucy** [reply]
Thanks for the help!

**Reply by Absolucy** [reply]
@khvzak wait, with that method, I always have to pass the userdata back into it self? like `router.register(router, "blah", function() end)`, even though I want there to just be one global instance per state.

**Reply by khvzak** [reply]
> @khvzak wait, with that method, I always have to pass the userdata back into it self? like `router.register(router, "blah", function() end)`, even though I want there to just be one global instance per state.

In Lua you could use a short form: `router:register(...)` which is equals to `router.register(router, ...)` (alternatively you could call `router.register()` and then fetch a global instance inside the method, but it's less efficient).

Regarding maintaining `callback_idx`, more efficient I think is to set a free slot to `Nil` and ideally tracking free slots in order to re-use them for further `register` calls. 

Sorry that `UserData` does not allow to store a reference to a Lua `Value<'lua>`. 
This is solvable by changing prototype of the `UserData` trait to:
```rust
pub trait UserData<'lua>: Sized {
    fn add_methods<M: UserDataMethods<'lua, Self>>(_methods: &mut M) {}
}
```
But this is a breaking change. I might consider adding a `UserDataRef<'lua>` trait to keep backward compatibility. Need to think about it.

**Reply by Absolucy** [reply]
Wait, how do I pull the `Router` out of the `&Lua` from elsewhere? Trying to `get_raw` it just complains that it needs `Clone`, although I can't find where `Clone` is required.

**Reply by khvzak** [reply]
If router is a Lua global object, you can get it via `lua.globals().get("router")?;` (as your original type or `AnyUserData`)

**Reply by Absolucy** [reply]
```
error[E0277]: the trait bound `Router: Clone` is not satisfied
  --> src/server/worker.rs:26:38
   |
26 |         let router: Router = lua.globals().get("router").unwrap();
   |                                            ^^^ the trait `Clone` is not implemented for `Router`
   |
   = note: required because of the requirements on the impl of `FromLua<'_>` for `Router`
```

Which doesn't make sense, because `FromLua<'lua>: Sized` and I don't think `Sized` needs `Clone` (it's the other way around)?

**Reply by khvzak** [reply]
`Clone` requirement is defined there: 
https://github.com/khvzak/mlua/blob/master/src/conversion.rs#L131
In this case you need to wrap `Router` to `Arc` (because there is no way to move value out from Lua in a Rust way).

Alternative way:
```rust
let router_ud = lua.globals().get::<_, AnyUserData>("router")?;
let router = router_ud.borrow::<Router>()?;
```
You can also use `router_ud` to get stored callbacks. `router` will have type `Ref<Router>` which is similar to `&Router`.
https://doc.rust-lang.org/std/cell/struct.Ref.html

**Reply by Absolucy** [reply]
@khvzak thanks, I was just using "Go to Definition" which took me to value.rs
